src/cargo/sources/path.rs

```diff
@@ -420,9 +421,23 @@ impl<'cfg> PathSource<'cfg> {
                 true
             });
         for entry in walkdir {
-            let entry = entry?;
-            if !entry.file_type().is_dir() {
-                ret.push(entry.path().to_path_buf());
+            use std::io::{Error, ErrorKind::PermissionDenied};
+            match entry {
+                Ok(entry) => {
+                    if !entry.file_type().is_dir() {
+                        ret.push(entry.into_path());
+                    }
+                }
+                Err(err) if err.loop_ancestor().is_some() => {
+                    self.config.shell().warn(err)?;
+                }
+                Err(err) => match (err.path(), err.io_error()) {
+                    (_, Some(err)) if err.kind() == PermissionDenied => {
+                        return Err(Error::new(PermissionDenied, err.to_string()).into());
+                    }
+                    (Some(path), _) => ret.push(path.to_path_buf()),
```

Comment:
I'm curious, what's going on here? This looks like it's ignoring all errors that are not permission denied?

Reply:
If cargo doesn't handle permission error intentionally, this [`EACCES` test case](https://github.com/rust-lang/cargo/blob/0.58.0/tests/testsuite/build_script.rs#L4623-L4654) might pass silently. The old behavior before adopting walkdir returns errors from  `fs::read_dir` calls directly, as well as it just unwraps `DirEntry` without handling `Result`.
https://github.com/rust-lang/cargo/blob/a359ce16073401f28b84840da85b268aa3d37c88/src/cargo/sources/path.rs#L413-L416

With walkdir, cargo cannot tell whether the error is permission error or a broken link. Generally, walkdir emits three kinds of errors:

1. **Symlink loop errors**: Cargo already takes care of it and shows a warning in this PR.
2. **IO errors with a path**: I guess cargo can recover from the path, and later let the caller decide to ignore it or if the caller does access the path and hits the error.
3. **IO errros without a path**: Cargo can simply propagate it to callers. This is rare because walkdir only emits this kind for either opening the file to check symlink loop or  handling the `Result<DirEntry>` generated from `read_dir()`.

I've updated this PR to reflect aforementioned change. There is an enhancement can do such as emitting warnings for those recoverable errors, but I doubt its usefulness.